pkg/security/resolvers/file/resolver_linux.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/DataDog/datadog-agent/pkg/security/utils"
+	"github.com/DataDog/datadog-agent/pkg/security/utils/lru"
+	"github.com/DataDog/datadog-go/v5/statsd"
+)
+
+const (
+	checkLinkage                  = false
+	fileMetadataResolverCacheSize = 512
+)
+
+// Opt defines options for resolvers
+type Opt struct {
+	CgroupResolver *cgroup.Resolver
+}
+
+// Resolver represents a cache for mountpoints and the corresponding file systems
+type Resolver struct {
+	cache *lru.Cache[LRUCacheKey, *model.FileMetadatas]
+
+	cgroupResolver *cgroup.Resolver
+
+	// stats
+	statsdClient statsd.ClientInterface
+	cacheHit     *atomic.Uint64
+	cacheMiss    *atomic.Uint64
+}
+
+// LRUCacheKey is the structure used to access cached metadatas
+type LRUCacheKey struct {
+	containerID containerutils.ContainerID
+	path        string
+	mTime       int64
+}
+
+// NewResolver returns a new instance of the hash resolver
+func NewResolver(statsdClient statsd.ClientInterface, opt *Opt) (*Resolver, error) {
+	cache, err := lru.New[LRUCacheKey, *model.FileMetadatas](fileMetadataResolverCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create file metadatas resolver cache: %w", err)
+	}
+
+	return &Resolver{
+		statsdClient:   statsdClient,
+		cgroupResolver: opt.CgroupResolver,
+		cache:          cache,
+		cacheHit:       atomic.NewUint64(0),
+		cacheMiss:      atomic.NewUint64(0),
+	}, nil
+}
+
+// ResolveFileMetadatas resolves file metadatas
+func (r *Resolver) ResolveFileMetadatas(event *model.Event, file *model.FileEvent) (*model.FileMetadatas, error) {
+	if !file.IsPathnameStrResolved {
+		event.FieldHandlers.ResolveFilePath(event, file)
+	}
+	if file.PathResolutionError != nil {
+		return nil, errors.New("path resolution error")
+	}
+
+	// fileless
+	if file.BasenameStr == "memfd:" && file.PathnameStr == "" {
+		return &model.FileMetadatas{
+			Type:         int(model.FileLess),
+			IsExecutable: true,
+		}, nil
+	}
+
+	// add pid one for hash resolution outside of a container
+	process := event.ProcessContext.Process
+	rootPIDs := []uint32{process.Pid, 1}
+	if event.ProcessCacheEntry != nil {
+		rootPIDs = event.ProcessCacheEntry.GetAncestorsPIDs()
+	}
+	if process.ContainerID != "" && r.cgroupResolver != nil {
+		w, ok := r.cgroupResolver.GetWorkload(process.ContainerID)
+		if ok {
+			rootPIDs = w.GetPIDs()
+		}
+	}
+
+	for _, pid := range rootPIDs {
+		//  get proc path
+		path := utils.ProcRootFilePath(pid, file.PathnameStr)
+
+		// get file stats
+		fileInfo, err := os.Stat(path)
+		if err != nil {
+			continue
+		}
+
+		// init the cache key to lookup
+		key := LRUCacheKey{
+			containerID: process.ContainerID,
+			path:        path,
+			mTime:       fileInfo.ModTime().UnixNano(),
+		}
+
+		// cache lookup
+		entry, ok := r.cache.Get(key)
+		if ok {
+			r.cacheHit.Inc()
+			return entry, nil
+		}
+
+		// if no result in cache, analyze the file an return result
+		info, err := AnalyzeFile(path, fileInfo, checkLinkage)
```

Comment:
You need to check that this is a regular file. Named pipe for instance will block